src/Npgsql/NpgsqlReadBuffer.cs

```diff
@@ -185,36 +188,64 @@ async Task EnsureLong()
+                        // Stopping twice (in case the previous Stop() call succeeded) doesn't hurt.
+                        // Not stopping will cause an assertion failure in debug mode when we call Start() the next time.
+                        // We can't stop in a finally block because Connector.Break() will dispose the buffer and the contained
+                        // _timeoutCts
+                        Cts.Stop();
+                        switch (e)
+                        {
+                        // User requested the cancellation
+                        case OperationCanceledException _ when (cancellationToken.IsCancellationRequested):
+                            throw readingNotifications ? e : Connector.Break(e);
+                        // Read timeout
+                        case OperationCanceledException _:
+                        // Note that mono throws SocketException with the wrong error (see #1330)
+                        case IOException _ when (e.InnerException as SocketException)?.SocketErrorCode ==
+                                                (Type.GetType("Mono.Runtime") == null ? SocketError.TimedOut : SocketError.WouldBlock):
+                            Debug.Assert(e is OperationCanceledException ? async : !async);
+
+                            if (!isCancellation && !readingNotifications)
+                            {
+                                var skipCancellationWait = false;
+
+                                try
+                                {
+                                    Connector.CancelRequest(throwExceptions: true, requestedByUser: false);
+                                }
+                                catch
+                                {
+                                    skipCancellationWait = true;
+                                }
+
+                                if (!skipCancellationWait)
+                                {
+                                    isCancellation = true;
+
+                                    if (Connector.Settings.CancellationTimeout > 0)
+                                        Timeout = TimeSpan.FromMilliseconds(Connector.Settings.CancellationTimeout);
+
+                                    if (async)
+                                        finalCt = Cts.Start(cancellationToken);
+
+                                    continue;
+                                }
+                            }
+
+                            var timeoutException = new NpgsqlException("Exception while reading from stream", new TimeoutException("Timeout during reading attempt"));
+                            throw readingNotifications ? timeoutException : Connector.Break(timeoutException);
+                        }
+
+                        throw Connector.Break(new NpgsqlException("Exception while reading from stream", e));
```

Comment:
nit: this can move into a default label in the switch, to slightly improve readability